db/types.go

```diff
@@ -0,0 +1,171 @@
+	// Returns nil for invalid versions.
+	ReaderAt(uint64) DBReader
+
+	// Opens a read-write transaction at the current version.
+	ReadWriter() DBReadWriter
+
+	// Opens a write-only transaction at the current version.
+	Writer() DBWriter
+
+	// Returns all saved versions
+	Versions() VersionSet
+
+	// Saves the current version of the database and returns its ID.
+	// Waits for any pending Writer transactions to be discarded. TODO: is an error preferred?
+	SaveVersion() uint64
+
+	// Close closes the database connection.
+	Close() error
+
+	// Stats returns a map of property values for all keys and the size of the cache.
+	Stats() map[string]string
+}
+
+// TODO: rename to batch/transaction?
+// The main DB access interface. Follows BadgerDB semantics , batches are
+// concurrency-safe, and produce errors on write conflicts.
+// Callers must call Close on the batch when done.
+//
+// Keys cannot be nil or empty, while values cannot be nil. Keys and values should be considered
+// read-only, both when returned and when given, and must be copied before they are modified.
+type DBReader interface {
+	// Get fetches the value of the given key, or nil if it does not exist.
+	// CONTRACT: key, value readonly []byte
+	Get([]byte) ([]byte, error)
+
+	// Has checks if a key exists.
+	// CONTRACT: key, value readonly []byte
+	Has(key []byte) (bool, error)
+
+	// Iterator returns an iterator over a domain of keys, in ascending order. The caller must call
+	// Close when done. End is exclusive, and start must be less than end. A nil start iterates
+	// from the first key, and a nil end iterates to the last key (inclusive). Empty keys are not
+	// valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
+
+	// Discards the transaction, invalidating any future operations on it.
+	Discard()
+}
+
+// A write-only batch; can be used to wrap write-optimized batches
+type DBWriter interface {
+	// Set sets the value for the given key, replacing it if it already exists.
+	// CONTRACT: key, value readonly []byte
+	Set([]byte, []byte) error
+
+	// Delete deletes the key, or does nothing if the key does not exist.
+	// CONTRACT: key readonly []byte
+	Delete([]byte) error
+
+	// Flushes pending writes and discards the transaction.
+	// TODO: maybe change to Flush() and follow WriteBatch semantics (ie. don't discard)
+	Commit() error
```

Comment:
I prefer `Commit` than `Flush`.

Reply:
I do too, but Flush may be more appropriate if the batch/txn were reusable IMO

Reply:
Even fro batch transactions, Commit is more meaningful than flush. In general, when we talk about transactions we want to commit rather than flush ;)